src/liballoc/boxed.rs

```diff
@@ -63,6 +63,28 @@
 //! T` obtained from `Box::<T>::into_raw` may be deallocated using the
 //! [`Global`] allocator with `Layout::for_value(&*value)`.
 //!
+//! `Box<T>` has the same representation as `*mut T`. In particular, when
```

Comment:
Is "the same representation" precise enough? Size and alignment are the same, but the niche is not. Producing a NULL `Box<T>` is insta-UB, unlike for `*mut T`.

Reply:
The size, alignment, niches, call ABI, and value representation relation of `Box<T>` match that of `&mut T`.

The validity invariant of `Box` might match that of `&mut T` as well, see the UCG issue for the [validity of Box](https://github.com/rust-lang/unsafe-code-guidelines/issues/145). 

I'm not sure how relevant it is that `Box` has a pointer field and `&mut T` has no fields, given that their call ABI is the same and both are Scalar::Pointer.

Reply:
Wait, the whole point of this change is to guarantee that Box can be used in C FFI. 

Guaranteeing that `Box<T>` has the same _call ABI_ as `*mut T` would be enough to solve that problem.

Reply:
> Guaranteeing that Box<T> has the same call ABI as *mut T would be enough to solve that problem.

Well for passing things in-memory it also needs to have the same "memory ABI" if that's a thing.

Reply:
Do we have similar wording for references anywhere? We should probably sync that up. But I couldn't fit it in the [reference type docs](https://doc.rust-lang.org/nightly/std/primitive.reference.html)?

Reply:
> Well for passing things in-memory it also needs to have the same "memory ABI" if that's a thing.

True. At this point we might just want to say that `Box<T>` has the same ABI as a `&mut T` (not just call ABI, but niches, and anything else that might matter).

Reply:
> we might just want to say that Box<T> has the same ABI as a &mut T

Yeah, I like that. Everything should be equal for those two.

Then, can we link to some other document talking about the use of references for FFI?

Reply:
The API docs of [references](https://doc.rust-lang.org/std/primitive.reference.html) mention, e.g., 

> In fact, Option<&T> has the same memory representation as a nullable pointer, and can be passed across FFI boundaries as such.

but that's all I've been able to find written down anywhere :/

Reply:
IMO the only difference between `Box<T>` and `&'static mut T` (or more accurately, `&'a mut T` where `T: 'a` by virtue of `'a` being the shortest lifetime in `T`, so sort of `for<'a> &'a mut T`) is the set of operations you can do to it, which is strictly larger for `Box<T>` (`Box::leak` effectively letting you get the `&mut T`, but nothing safe goes in the other direction).

So, yeah, I wouldn't mind guaranteeing that they have identical ABI.

Reply:
@stephaneyfx Want to update the PR based on the conversation ^---? :)

Reply:
@Centril Sorry for the delay. I will update the PR this week.

Reply:
No problem. Please take your time. :)